tests/mocha/test_helpers.js

```diff
@@ -42,13 +45,37 @@ function captureWarnings(innerFunc) {
   return msgs;
 }
 
+/**
+ * Creates stub for Blockly.utils.genUid that returns the provided id or ids.
+ *    Recommended to also assert that the stub is called the expected number of
+ *    times.
+ * @param {string|!Array<string>} returnIds The return values to use for the
+ *    created stub. If a single value is passed, then the stub always returns
+ *    that value.
+ * @return {!SinonStub} The created stub.
+ */
+function createGenUidStubWithReturns(returnIds) {
+  var stub = sinon.stub(Blockly.utils, "genUid");
+  if (Array.isArray(returnIds)) {
+    for (var i = 0; i < returnIds.length; i++) {
+      stub.onCall(i).returns(returnIds[i]);
+    }
+  } else {
+    stub.returns(returnIds);
+  }
+  return stub;
+}
 
 /**
  * Creates stub for Blockly.Events.fire that fires events immediately instead of
  * with timeout.
- * @return {sinon.stub} The created stub.
+ * @return {!SinonStub} The created stub.
  */
 function createEventsFireStub() {
+  // TODO(#4064): Remove clearing of event clear here in favor of adding cleanup
+  // to other tests that cause events to be added to the queue even after they
+  // end.
+  Blockly.Events.FIRE_QUEUE_.length = 0;
```

Comment:
This fix **seems** to work.

However, I **thought** I had tested with this change at some point and it didn't work, but it works now.
I'm a little worried that this fixes most failures, but there's still "sometimes" failures, which is why I want to address this better in a follow up as part of #4064.